### PR TITLE
added lesson creation flow

### PR DESCRIPTION
**Pull request by mazemaze**
Users can set their thumbnail and a preview image and video for a lesson. 
Could pick resource files from local storage and allow to pick pdf files.
Most of the codes came from the salon-creation-screen.

https://youtu.be/ox71j-NCJqs

### CODE REVIEW

File: salon_creator/lib/app.dart [hunk 1]
```diff
@@ -32,6 +33,7 @@ class SalonCreatorApp extends StatelessWidget {
         '/terms': (context) => TermsScreen(),
         '/user/profile/get': (context) => UserProfileScreen(),
         '/user/profile/update': (context) => UserProfileEditScreen(),
+        'lesson_creation': (context) => LessonCreationScreen(),
```
Comment (dragonfly91) [inline salon_creator/lib/app.dart]:
```suggestion
        '/lesson/create': (context) => LessonCreationScreen(),
```

This is more REST-like

File: salon_creator/lib/common/image_picker.dart [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import 'package:image_picker/image_picker.dart';
+
+Future<XFile> imagePicker() async {
```
Comment (dragonfly91) [inline salon_creator/lib/common/image_picker.dart]: I don't see the reason why this has to be in a separate function.

File: salon_creator/lib/firebase/database.dart [hunk 3]
```diff
@@ -41,24 +44,45 @@ class DbHandler {
       .get()
       .docs
       .first;
+
+  Future addLesson(Lesson lesson, String userId) async =>
+      await salonCollectionRef
+          .doc(userId)
```
Comment (dragonfly91) [inline salon_creator/lib/firebase/database.dart]: are you sure this is userId and not salonId? I assumed salons were indexed by a unique ID

File: salon_creator/lib/firebase/database.dart [hunk 4]
```diff
@@ -41,24 +44,45 @@ class DbHandler {
...
   }
 
+  Future<String> uploadResourceAndGetUrl(File file) async {
+    if (file != null) {
+      var snapshot =
+          await FirebaseStorage.instance.ref().child('resources').putFile(file);
+
+      return snapshot.ref.getDownloadURL();
+    }
+    return "";
```
Comment (dragonfly91) [inline salon_creator/lib/firebase/database.dart]: It is better to return null instead of an empty string

File: salon_creator/lib/models/lesson.dart [hunk 5]
```diff
@@ -0,0 +1,45 @@
+class Lesson {
+  String salonId;
+  String name;
+  String thumbnail;
```
Comment (dragonfly91) [inline salon_creator/lib/models/lesson.dart]: thumbnailUrl ?

File: salon_creator/lib/models/lesson.dart [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  String media;
+  String category;
+  String description;
+  bool publish;
```
Comment (dragonfly91) [inline salon_creator/lib/models/lesson.dart]: isPublished

File: salon_creator/lib/models/lesson.dart [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  List<String> resources;
```
Comment (dragonfly91) [inline salon_creator/lib/models/lesson.dart]: I think it is better to model this as `List<Resource>` where:

```
class Resource {
  String displayName;
  String url;
}
```

File: salon_creator/lib/screens/home.dart [hunk 8]
```diff
@@ -36,6 +36,15 @@ class _HomePageState extends State<HomePage> {
           length: 2,
           child: Scaffold(
             appBar: AppBar(
+              actions: [
+                IconButton(
+                  onPressed: () {
+                    Navigator.of(context).pushNamed('lesson_creation');
+                  },
+                  icon: Icon(Icons.add_circle_outline),
+                  color: primaryColor,
```
Comment (dragonfly91) [inline salon_creator/lib/screens/home.dart]: I think specifying this is not needed, as this is set in the theme. Did you try not specifying the color here?

File: salon_creator/lib/screens/lesson_creation_screen.dart [hunk 9]
```diff
@@ -0,0 +1,586 @@
...
+
+class LessonCreationScreen extends StatefulWidget {
+  const LessonCreationScreen({Key key}) : super(key: key);
+
+  @override
+  _LessonCreationScreenState createState() => _LessonCreationScreenState();
+}
+
+class _LessonCreationScreenState extends State<LessonCreationScreen> {
+  bool _publish = false;
```
Comment (dragonfly91) [inline salon_creator/lib/screens/lesson_creation_screen.dart]:
```suggestion
  bool _isPublished = false;
```

File: salon_creator/lib/screens/lesson_creation_screen.dart [hunk 10]
```diff
@@ -0,0 +1,586 @@
...
+  bool isFormFilled = false;
```
Comment (dragonfly91) [inline salon_creator/lib/screens/lesson_creation_screen.dart]: Either make everything private or everything public. I prefer private though

File: salon_creator/lib/screens/lesson_creation_screen.dart [hunk 11]
```diff
@@ -0,0 +1,586 @@
...
+      allowMultiple: true,
+      type: FileType.custom,
+      allowedExtensions: ['pdf'],
+    );
+    if (result != null) {
+      for (int i = 0; i < result.count; i++) {
+        setState(() {
+          resourcesList.add(result.names[i]);
+          resources.add(File(result.paths[i]));
+          print(resources[i]);
```
Comment (dragonfly91) [inline salon_creator/lib/screens/lesson_creation_screen.dart]: please remove print statements

File: salon_creator/lib/screens/lesson_creation_screen.dart [hunk 12]
```diff
@@ -0,0 +1,586 @@
...
+    final screenHeight = mq.height;
+    return SafeArea(
+      child: GestureDetector(
+        onTap: () {
+          FocusScope.of(context).unfocus();
+        },
+        child: SingleChildScrollView(
+          child: Container(
+            child: Column(
+              children: [
```
Comment (dragonfly91) [inline salon_creator/lib/screens/lesson_creation_screen.dart:133]: Really like this. Nice work! (self-resolve)

File: salon_creator/lib/screens/lesson_creation_screen.dart [hunk 13]
```diff
@@ -0,0 +1,586 @@
...
+  }
+
+  Widget _buildCreateInProgressScreen() {
+    return Center(
+      child: Column(
+        mainAxisAlignment: MainAxisAlignment.center,
+        children: [
+          Text(
+            "作成中です",
+            style: TextStyle(
```
Comment (dragonfly91) [inline salon_creator/lib/screens/lesson_creation_screen.dart]: Please use existing themes if possible

File: salon_creator/lib/screens/lesson_creation_screen.dart [hunk 14]
```diff
@@ -0,0 +1,586 @@
...
+  Widget _buildMediaPreviewerDivider(String text) {
+    return Row(
+      mainAxisAlignment: MainAxisAlignment.center,
+      children: [
+        Text(text),
+      ],
+    );
+  }
+
+  delete(index) {
```
Comment (dragonfly91) [inline salon_creator/lib/screens/lesson_creation_screen.dart]: This naming is too generic, please rename to something more specific

File: salon_creator/lib/screens/lesson_creation_screen.dart [hunk 15]
```diff
@@ -0,0 +1,586 @@
...
+          _buildResourceFileUploadForm(),
+          Divider(height: 32),
+          Row(
+            mainAxisAlignment: MainAxisAlignment.end,
+            children: [
+              Text(
+                "公開",
+                style: Theme.of(context).textTheme.headline3,
+              ),
+              Switch(value: _publish, onChanged: _changeSwtich),
```
Comment (dragonfly91) [inline salon_creator/lib/screens/lesson_creation_screen.dart]:
```suggestion
              Switch(value: _publish, onChanged: _changeSwitch),
```

File: salon_creator/lib/screens/lesson_creation_screen.dart [hunk 16]
```diff
@@ -0,0 +1,586 @@
...
+  Widget _buildResourceFileUploadForm() {
+    return Column(
+      children: [
+        Row(
+          mainAxisAlignment: MainAxisAlignment.spaceBetween,
+          children: [
+            CustomLabel(title: "リソース"),
+            ElevatedButton(
+              onPressed: () {
+                _filePicker();
```
Comment (dragonfly91) [inline salon_creator/lib/screens/lesson_creation_screen.dart]: Functions should be named with verbNoun convention. So ideally, this should be `_pickFile()`

File: salon_creator/lib/screens/lesson_creation_screen.dart [hunk 17]
```diff
@@ -0,0 +1,586 @@
...
+        Navigator.pop(context);
+      },
+      rightButtonText: "取り消し",
+      rightFunction: () {
+        Navigator.of(context).pop();
+      },
+    );
+  }
+
+  removeFile() {
```
Comment (dragonfly91) [inline salon_creator/lib/screens/lesson_creation_screen.dart]: why not make it private? Make all functions and variables private as much as possible, please.

File: salon_creator/lib/screens/lesson_creation_screen.dart [hunk 18]
```diff
@@ -0,0 +1,586 @@
...
+    final dbHandler = DbHandler();
+    final storageHandler = StorageHandler();
+    Lesson lesson = Lesson();
+    lesson.resources = [];
+    lesson.salonId = "XytREh97Xs9pJvnI13JS";
+    lesson.name = lessonNameController?.text;
+    lesson.category = categoriesController?.text;
+    lesson.description = descriptionController?.text;
+    lesson.publish = _publish;
+    lesson.media = _mediaFile != null
```
Comment (dragonfly91) [inline salon_creator/lib/screens/lesson_creation_screen.dart]:
```suggestion
    lesson.imageUrl = _mediaFile != null
```

**Review by dragonfly91 — CHANGES_REQUESTED** [review]
Please address comments and fix merge conflicts

File: salon_creator/lib/screens/lesson_creation_screen.dart (hunk 18 above)
Comment (mazemaze) [reply salon_creator/lib/screens/lesson_creation_screen.dart]: This could be either a image or video. So, it's probably better to name it as mediaUrl instead of imageUrl. But I'm not sure if it's proper to name it as mediaUrl.

File: salon_creator/lib/screens/home.dart (hunk 8 above)
Comment (mazemaze) [reply salon_creator/lib/screens/home.dart]: Well, it actually needs it 'cause without it the color of IconButton was black.
However, I've decided to remove it as a reason for matching the color of items in the appbar.

File: salon_creator/lib/screens/lesson_creation_screen.dart (hunk 14 above)
Comment (mazemaze) [reply salon_creator/lib/screens/lesson_creation_screen.dart]: I just forgot to delete this function

File: salon_creator/lib/screens/lesson_creation_screen.dart (hunk 17 above)
Comment (mazemaze) [reply salon_creator/lib/screens/lesson_creation_screen.dart]: I just forgot to delete this function

File: salon_creator/lib/models/lesson.dart [hunk 19]
```diff
@@ -1,45 +1,48 @@
...
   String name;
-  String thumbnail;
-  String media;
+  String thumbnailUrl;
+  String mediaUrl;
   String category;
   String description;
-  bool publish;
-  List<String> resources;
+  bool isPublish;
```
Comment (dragonfly91) [inline salon_creator/lib/models/lesson.dart:8]:
```suggestion
  bool isPublished;
```

Sorry for being super picky

File: salon_creator/lib/screens/lesson_creation_screen.dart [hunk 20]
```diff
@@ -21,32 +21,44 @@ class LessonCreationScreen extends StatefulWidget {
...
 class _LessonCreationScreenState extends State<LessonCreationScreen> {
-  bool _publish = false;
-  bool isFormFilled = false;
-  bool operationInProgress = false;
-  List<File> resources = [];
-  List<String> resourcesList = [];
-  TextEditingController categoriesController = TextEditingController();
-  TextEditingController descriptionController = TextEditingController();
-  TextEditingController lessonNameController = TextEditingController();
+  bool _isPublish = false;
```
Comment (dragonfly91) [inline salon_creator/lib/screens/lesson_creation_screen.dart]:
```suggestion
  bool _isPublished = false;
```

Sorry for being super picky

File: salon_creator/lib/screens/lesson_creation_screen.dart [hunk 21]
```diff
@@ -21,32 +21,44 @@ class LessonCreationScreen extends StatefulWidget {
...
+  TextEditingController _categoriesController = TextEditingController();
+  TextEditingController _descriptionController = TextEditingController();
+  TextEditingController _lessonNameController = TextEditingController();
   XFile _imageFile;
   XFile _mediaFile;
-  XFile thumbnailFile;
-  XFile image;
+  XFile _thumbnailFile;
+
+  Future<XFile> imagePicker() async {
```
Comment (dragonfly91) [inline salon_creator/lib/screens/lesson_creation_screen.dart]: Similar to the pickFile function, this could be better named:

```suggestion
  Future<XFile> _pickImage() async {
```

**Review by dragonfly91 — DISMISSED** [review]
Left a few minor comments. Please check!

File: salon_creator/lib/screens/home.dart [hunk 22]
```diff
@@ -34,7 +34,31 @@ class _HomePageState extends State<HomePage> {
 
         var creatorModel = CreatorModel.fromMap(snapshot.data.data());
 
-        return _buildHomePageInternal(context, creatorModel);
```
Comment (dragonfly91) [inline salon_creator/lib/screens/home.dart]: Please undo this change since we now have a nested tab controller

**Review by dragonfly91 — DISMISSED** [review]
Please address one comment

**Review by dragonfly91 — APPROVED** [review]
